### ISSUE TITLE
Undo tree display / navigation

### ISSUE THREAD

**Issue by haasn** [zed-industries/zed#17455]
### Check for existing issues

- [X] Completed

### Describe the feature

A very popular (and useful) feature among editors is the ability to maintain, display, and navigate an undo tree. This way you can e.g.:

- Quickly view a previous version of the buffer you're editing
- Make changes after undoing a (sequence of) commands without losing progress
- Inspect your history of saved versions and quickly navigate between older variants (e.g. for testing)
- Easily go back to the latest version of a buffer after a sequence of undo commands

And definitely more. In general, it gives us a way to track changes on a level of granularity too fine for e.g. git commits to sensibly track.

### If applicable, add mockups / screenshots to help present your vision of the feature

Vim's undotree plugin can offer some inspiration:

![image](https://github.com/user-attachments/assets/ad799f4b-e29b-42ea-aeca-5e7c78cf50d0)

(Observe also the fantastic utility here of having #15097]

**Comment by aeshna-cyanea** [comment]
for reference (to show that there is demand), here is an equivalent issue in vscode 
- https://github.com/microsoft/vscode/issues/20889

**Comment by leo-fresha** [comment]
I just lost a couple of hours of work due my messing up while doing a bunch of undoes and redos. This would've been a life saver.

**Comment by muja** [comment]
This is an absolute must for `Zed`, especially considering whenever you save the buffer, it will (by default) run the formatter, which will completely delete the entire Redo history, _even if the buffer is unchanged by the formatter_. This is one of the roughest edges around Zed right now.

**Comment by cies** [comment]
I agree this more important than ever, with AI agent-written changes becoming ever more common.

**Comment by SL-RU** [comment]
UNDO TREE in emacs must be a reference, this is the best undo tree implementation

**Comment by muja** [comment]
> Hi there! 👋 We're working to clean up our issue tracker by closing older bugs that might not be relevant anymore. If you are able to reproduce this issue in the latest version of Zed, please let us know by commenting on this issue, and it will be kept open. If you can't reproduce it, feel free to close the issue yourself. Otherwise, it will close automatically in 14 days. Thanks for your help!

Yes, this can be reproduced in the latest version of Zed:

1. open Zed
2. no undo tree

**Comment by miguelraz** [comment]
Sorry, the `stalebot` upstream went over eager - please ignore.

**Comment by GoldStrikeArch** [comment]
Might be implemented in this [PR](https://github.com/zed-industries/zed/pull/58002)
